akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala

```diff
@@ -200,6 +243,90 @@ private[akka] object Running {
+              this
+            case Some(replicaId) =>
+              onPublishedEvent(state, activeActive, replicaId, event)
+          }
+      }
+    }
+
+    private def onPublishedEvent(
+        state: Running.RunningState[S],
+        activeActive: ActiveActive,
+        originReplicaId: String,
+        event: PublishedEventImpl): Behavior[InternalProtocol] = {
+      val log = setup.log
+      val separatorIndex = event.persistenceId.id.indexOf(PersistenceId.DefaultSeparator)
+      val idPrefix = event.persistenceId.id.substring(0, separatorIndex)
+      if (!setup.persistenceId.id.startsWith(idPrefix)) {
+        log.warn("Ignoring published replicated event for the wrong actor [{}]", event.persistenceId)
+        this
+      } else if (originReplicaId == activeActive.replicaId) {
+        // FIXME MultiDC handled this differently not sure why we'd need our own published events?
+        if (log.isDebugEnabled)
+          log.debug(
+            "Ignoring published replicated event with seqNr [{}] from our own replica id [{}]",
+            event.sequenceNumber,
+            originReplicaId)
+        this
+      } else if (!activeActive.allReplicas.contains(originReplicaId)) {
+        log.warnN(
+          "Received published replicated event from replica [{}], which is unknown. Active active must be set up with a list of all replicas (known are [{}]).",
+          originReplicaId,
+          activeActive.allReplicas.mkString(", "))
+        this
+      } else {
+        val expectedSequenceNumber = state.seenPerReplica(originReplicaId) + 1
+        if (expectedSequenceNumber > event.sequenceNumber) {
+          // already seen
+          if (log.isDebugEnabled)
+            log.debugN(
+              "Ignoring published replicated event with seqNr [{}] from replica [{}] because it was already seen ([{}])",
+              event.sequenceNumber,
+              originReplicaId,
+              expectedSequenceNumber)
+          this
+        } else if (expectedSequenceNumber != event.sequenceNumber) {
+          //
+          if (log.isDebugEnabled) {
+            log.debugN(
+              "Ignoring published replicated event with replication seqNr [{}] from replica [{}] " +
+              "because expected replication seqNr was [{}] ",
+              event.sequenceNumber,
+              event.replicaId,
+              expectedSequenceNumber)
+          }
+          this
+        } else {
+          if (log.isTraceEnabled)
+            log.traceN(
+              "Received published replicated event [{}] with timestamp [{}] from replica [{}] seqNr [{}]",
+              Logging.simpleName(event.event.getClass),
+              Helpers.timestamp(event.timestamp),
+              originReplicaId,
+              event.sequenceNumber)
+
+          // fast forward stream for source replica
+          state.replicationControl.get(originReplicaId).foreach(_.fastForward(event.sequenceNumber))
```

Comment:
I think I've convinced my self there isn't a race here. We either get the ReplicatedEvent (RE) or the PublishedEvent first.

In this case we get the RE first we drop this event as the expected sequence nr is incorrect. 
If we get the PE first it will be delivered here.

Reply:
So the same replicated event could end up in our mailbox as the replication stream might get it through before this fast forward.

But the handling of a replicated event filters based on the seenPerReplica so I think that is okay.

In that case do we even need the fastForwawrd? I guess it avoids the ask between the replication stream and the ESB

Reply:
Yep. Fast forward optimizes away some extra work, especially if we have support for actual ffwd in journal and can avoid even reading anything from db.